### PR TITLE
Made rlog more testable and added unit tests

### PR DESCRIPTION
**Pull request by jbrendel**

### CODE REVIEW

File: rlog_test.go [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+var removeLogfile bool = false
+var fixedLogfileName bool = true
+
+// setup is called at the start of each test and prepares a new log file. It
+// also returns a new configuration, which can be used by this test.
+func setup() rlogConfig {
+	if fixedLogfileName {
+		logfile = "/tmp/rlog-test.log"
+	} else {
+		logfile = fmt.Sprintf("/tmp/rlog-test-%d.log", time.Now().UnixNano())
```
Comment (pritesh) [inline rlog_test.go:42]: fixedLogfileName is always true, so not sure why timestamp based one is there, since it can't be deleted using this method.

File: rlog_test.go [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+	}
+	os.Remove(logfile)
```
Comment (pritesh) [inline rlog_test.go:46]: why remove it in setup ?

File: rlog_test.go [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+		logFile:        logfile,
+		logStream:      "NONE",
+		logTimeDate:    false,
+		showCallerInfo: false,
+	}
+}
+
+// cleanup is called at the end of each test.
+func cleanup() {
+	if removeLogfile {
```
Comment (pritesh) [inline rlog_test.go:61]: removeLogfile is always false? so clean up never happens.

File: rlog_test.go [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+	conf.logLevel = "DEBUG"
+	conf.traceLevel = "1"
+	Initialize(conf)
+
+	Debugf("Test Debug %d", 123)
+	Infof("Test Info %d", 123)
+	Warnf("Test Warning %d", 123)
+	Errorf("Test Error %d", 123)
+	Criticalf("Test Critical %d", 123)
+	Tracef(1, "Trace 1 %d", 123)
```
Comment (pritesh) [inline rlog_test.go:184]: adding trace(2) here would help in negative testing, above negative is for trace, this would be for tracef.

File: rlog_test.go [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+	scanner := bufio.NewScanner(file)
+	i := 0
+	for scanner.Scan() {
+		line := scanner.Text()
+		if timeLayout != "" {
+			dateTime := line[:len(timeLayout)]
+			line = line[len(timeLayout)+1:]
+			_, err := time.Parse(timeLayout, dateTime)
+			if err != nil {
+				t.Fatal(err)
```
Comment (pritesh) [inline rlog_test.go]: this should be t.Error, can't have 2 fatals in row, fatal = error+stop

File: rlog_test.go [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+	// display
+	conf := setup()
+	defer cleanup()
+
+	conf.showCallerInfo = true
+	Initialize(conf)
+
+	Info("Test Info")
+	pc, fullFilePath, line, _ := runtime.Caller(0)
+	line-- // The log was called in the line before, so... -1
```
Comment (pritesh) [inline rlog_test.go:254]: nitpick: comment above line--, not on same one, kinda confusing becuase next line is different comment.

File: rlog_test.go [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+	Critical("Test Critical")
+
+	checkLines := []string{
+		"DEBUG    : Test Debug",
+		"INFO     : Test Info",
+		"WARN     : Test Warning",
+		"ERROR    : Test Error",
+		"CRITICAL : Test Critical",
+	}
+	fileMatch(t, checkLines, "")
```
Comment (pritesh) [inline rlog_test.go:140]: t.Log(checkLines) should help when using go test -v

File: rlog_test.go [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+	Trace(4, "Trace 4")
+	checkLines := []string{
+		"WARN     : Test Warning",
+		"ERROR    : Test Error",
+		"CRITICAL : Test Critical",
+		"TRACE(1) : Trace 1",
+		"TRACE(2) : Trace 2",
+		"TRACE(3) : Trace 3",
+	}
+	fileMatch(t, checkLines, "")
```
Comment (pritesh) [inline rlog_test.go:168]: ditto

**Review by pritesh — COMMENTED** [review]
rest looks good.

File: rlog.go [hunk 9]
```diff
@@ -209,49 +235,44 @@ func (f filter) match(filename string, level int) (bool, bool) {
...
 	settingDateTimeFormat = ""
-	if logTimeDate {
+	if config.logTimeDate {
 		// Store the format string for date/time logging. Allowed values are
 		// all the constants specified in
 		// https://golang.org/src/time/format.go.
 		var f string
-		switch logTimeFormatEnv {
+		switch config.logTimeFormat {
 		case "ANSIC":
```
Comment (debedb) [inline rlog.go:276]: would be nice if this was case-insensitive.

File: rlog_test.go (hunk 3 above)
Comment (debedb) [reply rlog_test.go:61]: I think the idea is that the test can set it to true (and also change fixedLogfileName as above)

File: rlog.go (hunk 9 above)
Comment (jbrendel) [reply rlog.go:276]: I can make this comparison here case insensitive, ok. The general format string, however, has to remain case sensitive, since people can define custom formats.

File: rlog_test.go (hunk 1 above)
Comment (jbrendel) [reply rlog_test.go:42]: I'd added the 'fixedLogfileName' flag as a means to test my own tests during development. In fact, I forgot to set this back to 'false'. I'll also add a comment for those flags, explaining that this is for internal testing. Otherwise, I'd have to edit lines of code in my tests while developing.

File: rlog_test.go (hunk 2 above)
Comment (jbrendel) [reply rlog_test.go:46]: I didn't want to add to the file, I want to make sure I'm always starting from scratch.

File: rlog_test.go (hunk 3 above)
Comment (jbrendel) [reply rlog_test.go:61]: Yes, what Greg said.

File: rlog_test.go (hunk 5 above)
Comment (jbrendel) [reply rlog_test.go]: That was a leftover from development. I'll remove that line.

File: rlog_test.go (hunk 7 above)
Comment (jbrendel) [reply rlog_test.go:140]: Ah, that's how I can get output during tests? Drove me nuts that I couldn't just print things! :-)

File: rlog_test.go (hunk 4 above)
Comment (jbrendel) [reply rlog_test.go:184]: Good point.

File: rlog_test.go (hunk 6 above)
Comment (jbrendel) [reply rlog_test.go:254]: Hm. Maybe.

File: rlog.go (hunk 9 above)
Comment (jbrendel) [reply rlog.go:276]: Done

File: rlog_test.go (hunk 1 above)
Comment (jbrendel) [reply rlog_test.go:42]: Added a comment.

File: rlog_test.go (hunk 2 above)
Comment (jbrendel) [reply rlog_test.go:46]: Added a comment.

File: rlog_test.go (hunk 7 above)
Comment (jbrendel) [reply rlog_test.go:140]: I've added a line by line 'should/is' output in fileMatch itself.

File: rlog_test.go (hunk 4 above)
Comment (jbrendel) [reply rlog_test.go:184]: Done.

File: rlog_test.go (hunk 6 above)
Comment (jbrendel) [reply rlog_test.go:254]: Added empty line after "line--" to make the comments more clear.

**Review by jbrendel — COMMENTED** [review]
Addressed review comments.
